lib/Backend/BackendUtils.cpp

```diff
@@ -123,6 +131,96 @@ bool isInput(const Placeholder *PH) {
   return false;
 }
 
+bool isOutput(const Placeholder *PH, const IRFunction &F) {
+  auto *weight = F.getWeightForNode(PH);
+  assert(weight && "Weight for a node was not found");
+  for (const auto &use : ValueUses(weight)) {
+    Instruction *user = use.get();
+    // Ignore deallocs.
+    if (isa<DeallocActivationInst>(user)) {
+      continue;
+    }
+    OperandKind kind = use.getOperand().second;
+    if (kind == OperandKind::Out || kind == OperandKind::InOut) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool isInput(const Placeholder *PH, const IRFunction &F) {
+  // Check that the PH is always used as an @in parameter by the current
+  // function.
+  auto *weight = F.getWeightForNode(PH);
+  assert(weight && "Weight for a node was not found");
+  for (const auto &use : ValueUses(weight)) {
+    Instruction *user = use.get();
+    // Ignore deallocs.
+    if (isa<DeallocActivationInst>(user)) {
+      continue;
+    }
+    OperandKind kind = use.getOperand().second;
+    if (kind == OperandKind::In || kind == OperandKind::InOut) {
+      return true;
+    }
+  }
+  return false;
+}
+
+template <typename FUN>
+std::vector<std::tuple<const Placeholder *, bool, bool>>
+getContiguousPlaceHolder(std::vector<const Placeholder *> const &holders,
+                         const FUN &F) {
+  // Pure input placeholders.
+  std::vector<const Placeholder *> intputPlaceholders;
+  // Pure output placeholders.
+  std::vector<const Placeholder *> outputPlaceholders;
+  // Input&output placeholders.
+  std::vector<const Placeholder *> inputOutputPlaceholders;
+  // Neither input nor output placeholders.
+  std::vector<const Placeholder *> emptyPlaceholders;
+  // return value.
+  std::vector<std::tuple<const Placeholder *, bool, bool>> ret;
+
+  for (auto &v : holders) {
+    if (isInput(v, F)) {
+      if (!isOutput(v, F)) {
+        intputPlaceholders.push_back(v);
+      } else {
+        inputOutputPlaceholders.push_back(v);
+      }
+    } else {
+      if (isOutput(v, F)) {
+        outputPlaceholders.push_back(v);
+      } else {
+        // Assert this empty may fail some test. Currently, we bypass the
+        // allocations. Need to log them somewhere else.
+        emptyPlaceholders.push_back(v);
```

Comment:
It would be nice to understand what kind of placeholders are in this category. If they are not read or written, are they used at all in the current function?

May be it is fixed now, after you fixed isInput/isOutput functions?

Reply:
> It would be nice to understand what kind of placeholders are in this category. If they are not read or written, are they used at all in the current function?
> 
> May be it is fixed now, after you fixed isInput/isOutput functions?

If we remove the allocation of the emptyPlaceholders, it still pass the tests. But assert it being empty will not.  I tried to look into one of the failed test, 
in the BackendCorrectnessTest.convGradTest, I dumped out the emptyPlaceholders. It has only one nodes named var2  which is connected to the selected of SoftMax and SoftMaxGrad function. Below is the DAG of the function after differentiation.
![conv1_F_after](https://user-images.githubusercontent.com/50883732/59174047-7d4f7480-8b04-11e9-9e1e-fcc776643eba.png)

Reply:
OK. Let's leave it for now then.